zoo_service.py

```diff
@@ -0,0 +1,107 @@
+db = SQLAlchemy(app)
+ma = Marshmallow(app)
+
+
+class Fish(db.Model):
+    id = db.Column(db.Integer, primary_key=True)
+    weight_in_kg = db.Column(db.Float)
+    thermoregulation = db.Column(db.String(15))
+    lifetime_years = db.Column(db.Float)
+    animal_type = db.Column(db.String(50))
+    required_aquarium_capacity_liters = db.Column(db.Float)
+    required_temperature = db.Column(db.Integer)
+    required_lighting_level = db.Column(db.String(20))
+
+    def __init__(
+            self, weight_in_kg, thermoregulation, lifetime_years, animal_type,
+            required_aquarium_capacity_liters, required_temperature, required_lighting_level
+    ):
+        self.weight_in_kg = weight_in_kg
+        self.thermoregulation = thermoregulation
+        self.lifetime_years = lifetime_years
+        self.animal_type = animal_type
+        self.required_aquarium_capacity_liters = required_aquarium_capacity_liters
+        self.required_temperature = required_temperature
+        self.required_lighting_level = required_lighting_level
+
+
+class FishSchema(ma.Schema):
+    id = fields.Int()
+    weight_in_kg = fields.Float()
+    thermoregulation = fields.String()
+    lifetime_years = fields.Float()
+    animal_type = fields.String()
+    required_aquarium_capacity_liters = fields.Integer()
+    required_temperature = fields.Integer()
+    required_lighting_level = fields.String()
+
+
+fish_schema = FishSchema(exclude=["id"])
+fishes_schema = FishSchema(only=["id", "animal_type", "lifetime_years"], many=True)
+
+
+@app.route('/')
+def hello_page():
+    return 'Hello to the zoo aquariums!'
+
+
+@app.route('/fish', methods=['GET'])
+def get_all_fish():
+    all_fish = Fish.query.all()
+    return fishes_schema.jsonify(all_fish)
+
+
+@app.route('/fish/<id>', methods=['GET'])
+def get_fish(id):
+    fish = Fish.query.get(id)
+    if fish is None:
+        return abort(404)
+    return fish_schema.jsonify(fish)
+
+
+@app.route('/fish', methods=['POST'])
+def add_fish():
+    try:
+        params = fish_schema.load(request.json)
+        new_fish = Fish(**params)
+    except:
+        return abort(400)
+    db.session.add(new_fish)
+    db.session.commit()
+    return fish_schema.jsonify(new_fish), 201
+
+
+@app.route('/fish/<id>', methods=['PUT'])
+def update_fish(id):
+    fish = Fish.query.get(id)
+    if fish is None:
+        return abort(404)
+    try:
+        new_params = fish_schema.load(request.json)
+    except ValidationError:
+        return abort(400)
+    for param in new_params:
+        setattr(fish, param, request.json[param])
+    db.session.commit()
+    return fish_schema.jsonify(fish)
+
+
+@app.route('/fish/<id>', methods=['DELETE'])
+def delete_fish(id):
+    fish = Fish.query.get(id)
+    if fish is None:
+        return abort(404)
+    db.session.delete(fish)
+    db.session.commit()
+    return fish_schema.jsonify(fish), 204
```

Comment:
HTTP204 is hardly a good idea here.

Reply:
I have read in the REST documentation that the DELETE request should return status code 204.

Reply:
Could you link that?
AFAIK there's no such thing as "REST documentation".

Reply:
I see.
When in doubt I'd follow the [RFC](https://tools.ietf.org/html/rfc7231#section-6.3.5).
HTTP204 technically _is_ acceptable here, just because it's a success response and typically you don't return a response body for DELETE requests. HTTP204 had been intended for a different use case (see the RFC), and I wouldn't expect to receive 204 in response to a DELETE request.

If you must use it, consider following the RFC to the letter and returning an empty response body.

Generally, a good practice would be to follow the principle of least surprise (google it!): use the broadest techniques (status codes etc) whenever possible, narrowing down where absolutely necessary. HTTP 400 and 404 is an important distinction, a client can be expected to do different things in response to 400 as opposed to 404; HTTP 200 and 204 are semantically almost the same.

Reply:
OK, thank you for all the information provided. Then I will change returned HTTP response to 200.  Also, if I understood correctly, should my PUT request return HTTP 204, or in this case, the response should be 200 too?